src/service/viewer-impl.js

```diff
@@ -998,92 +1025,87 @@ export class Viewer {
    * @param {string} eventType
    * @param {*} data
    * @param {boolean} awaitResponse
-   * @return {!Promise<*>|undefined}
+   * @return {!Promise<*>|undefined} the response promise if awaitResponse is
+   *     true, otherwise undefined
    */
   sendMessage(eventType, data, awaitResponse) {
     if (!this.messagingReadyPromise_) {
       return Promise.reject(getChannelError());
     }
     return this.messagingReadyPromise_.then(() => {
-      return this.sendMessageUnreliable_(eventType, data, awaitResponse);
+      return this.messageDeliverer_(eventType, data, awaitResponse);
     });
   }
 
   /**
-   * Broadcasts a message to all other AMP documents under the same viewer. It
-   * will attempt to deliver messages when the messaging channel has been
-   * established, but it will not fail if the channel is timed out.
+   * Sends the message to the viewer. This method queues up the message if the
+   * communication channel isn't established yet. The message will cancel an
+   * unsent message of the same type if seen in the queue.
    *
-   * @param {!JSONType} message
-   */
-  broadcast(message) {
-    this.maybeSendMessage_('broadcast', message);
-  }
-
-  /**
-   * Registers receiver for the broadcast events.
-   * @param {function(!JSONType)} handler
-   * @return {!UnlistenDef}
-   */
-  onBroadcast(handler) {
-    return this.broadcastObservable_.add(handler);
-  }
-
-  /**
-   * This message queues up the message to be sent when communication channel
-   * is established. If the communication channel is not established at
-   * this time, this method responds immediately with a Promise that yields
-   * `undefined` value.
    * @param {string} eventType
    * @param {*} data
    * @param {boolean} awaitResponse
-   * @return {?Promise<*>|undefined}
-   * @private
+   * @return {?Promise<*>|undefined} the response promise if awaitResponse is
+   *     true, otherwise undefined
    */
-  sendMessageUnreliable_(eventType, data, awaitResponse) {
+  sendMessageCancelUnsent(eventType, data, awaitResponse) {
     if (this.messageDeliverer_) {
       return this.messageDeliverer_(eventType, data, awaitResponse);
     }
 
-    // Store only a last version for an event type.
-    let found = null;
-    for (let i = 0; i < this.messageQueue_.length; i++) {
-      if (this.messageQueue_[i].eventType == eventType) {
-        found = this.messageQueue_[i];
-        break;
-      }
-    }
-    if (found) {
-      found.data = data;
+    const found = findIndex(this.messageQueue_, m => m.eventType == eventType);
+
+    let message;
+    if (found != -1) {
+      message = this.messageQueue_.splice(found, 1)[0];
```

Comment:
Oh, because we want the new message to be at the tail. Order matters here, think about message like `flushTick`.